lib/cloudant.js

```diff
@@ -604,29 +604,39 @@ Cloudant.prototype.replaceOrCreate = function replaceOrCreate(
   var self = this;
   var idName = self.idName(model);
   var mo = self.selectModel(model);
-
   var id = data[idName];
-  if (id) {
-    self.getCurrentRevision(model, id, function(err, rev) {
-      if (err) return cb(err);
-      if (rev) data._rev = rev;
-      self._insert(model, data, function(err) {
+
+  // Check if data exist first to avoid
+  // running getCurrentRevision() again an nonexistent instance
+  var where = {}; where[idName] = id;
+  var dataExist;
+
+  self.count(model, where, {}, function(err, cnt) {
+    if (err) return cb(err);
+    dataExist = cnt === 1 ? true : false;
+
+    if (id && dataExist === true) {
+      self.getCurrentRevision(model, id, function(err, rev) {
         if (err) return cb(err);
-        mo.db.get(id, function(err, doc) {
+        if (rev) data._rev = rev;
+        self._insert(model, data, function(err) {
           if (err) return cb(err);
-          cb(err, self.fromDB(model, mo, doc), {isNewInstance: !rev});
+          mo.db.get(id, function(err, doc) {
+            if (err) return cb(err);
+            cb(err, self.fromDB(model, mo, doc), {isNewInstance: !rev});
```

Comment:
I think in this branch `rev` always exists and `!rev` would always be `false`, right? In this case wouldn't it be more clear if you pass `{isNewInstance: false}` instead? Is my understanding right here? Correct me if I'm wrong.